stan/math/prim/fun/cholesky_corr_constrain.hpp

```diff
@@ -12,20 +12,22 @@
 namespace stan {
 namespace math {
 
-template <typename T>
-Eigen::Matrix<T, Eigen::Dynamic, Eigen::Dynamic> cholesky_corr_constrain(
-    const Eigen::Matrix<T, Eigen::Dynamic, 1>& y, int K) {
+template <typename T, require_t<is_eigen_vector<T>>* = nullptr>
+Eigen::Matrix<value_type_t<T>, Eigen::Dynamic, Eigen::Dynamic>
+cholesky_corr_constrain(const T& y, int K) {
   using Eigen::Dynamic;
   using Eigen::Matrix;
   using std::sqrt;
+  using T_scalar = value_type_t<T>;
   int k_choose_2 = (K * (K - 1)) / 2;
   check_size_match("cholesky_corr_constrain", "y.size()", y.size(),
                    "k_choose_2", k_choose_2);
-  Matrix<T, Dynamic, 1> z(k_choose_2);
+  Matrix<T_scalar, Dynamic, 1> z(k_choose_2);
+  const Eigen::Ref<const plain_type_t<T>>& y_ref = y;
   for (int i = 0; i < k_choose_2; ++i) {
-    z(i) = corr_constrain(y(i));
+    z[i] = corr_constrain(y_ref[i]);
   }
```

Comment:
Could we remove the loop and `Eigen::Ref` here through

```cpp
Matrix<T_scalar, Dynamic, 1> z = corr_constrain(y.head(k_choose_2));
```

with

```cpp
template <typename T>
inline auto corr_constrain(T&& x) {
  using std::tanh;
  return tanh(std::forward<T>(x));
}
```

Reply:
done